### ISSUE TITLE
clang analyzer: False positive NewDeleteLeaks for QObject related object creation

### ISSUE THREAD

**Issue by timxx** [llvm/llvm-project#56210]
See [Object Trees & Owership](https://doc.qt.io/qt-5/objecttrees.html), for example, the following code generates a false `clang-analyzer-cplusplus.NewDeleteLeaks`
``` c++
class Foo : public QObject {
public:
    Foo(QObject* parent) : QObject(parent) {
        QObject* bar = new QObject(this);
    }
};
```
Is it anyway to suppress this false report? Or how can I modify the mallocchecker to temporary disable this.

**Comment by martong** [comment]
I am wondering if you still have the false report if the definition of [QObjectPrivate::deleteChildren()]([QObjectPrivate](https://code.woboq.org/qt5/qtbase/src/corelib/kernel/qobject_p.h.html#QObjectPrivate)::deleteChildren()) is seen by the analyzer.
That function is the one that actually calls the delete. But that is implemented in `qobject.cpp` that is a separate TU. You may try and check whether you get the false report with [CTU](https://clang.llvm.org/docs/analyzer/user-docs/CrossTranslationUnit.html#automated-ctu-analysis-with-codechecker).

Not considering CTU, perhaps the `ownership_takes` attribute on the destructor of QObject could solve your issue. Could you try to annotate its prototype in your TU [like this](https://github.com/llvm/llvm-project/blob/5cbe39ef880e0d8a61db111fca0b14b3ca17a17b/clang/test/Analysis/malloc-annotations.cpp#L14)?

**Comment by timxx** [comment]
> Not considering CTU, perhaps the `ownership_takes` attribute on the destructor of QObject could solve your issue. Could you try to annotate its prototype in your TU [like this](https://github.com/llvm/llvm-project/blob/5cbe39ef880e0d8a61db111fca0b14b3ca17a17b/clang/test/Analysis/malloc-annotations.cpp#L14)?

Did you mean add the `ownership_takes` to the QObject destructor? Could you point me out where to find the documentation for ownership_takes? I don't see one on the internet.

**Comment by haoNoQ** [comment]
As long as you're not compiling Qt libraries yourself together with your project, CTU probably won't help.

This is definitely a problem for us, if we don't know what a function does we have to make judgement calls, and it's either this or we won't find any leaks at all (assuming *any* unknown constructor can do this). So we're sticking to the current behavior, seeing that these false positives are relatively rare.

Attribute `ownership_takes` can't be applied to `this`, and if it was, it would still not be a right solution because that'd cause us to have use-after-free warnings whenever the object is used.

I think a brand-new attribute could (and probably should) be built to cover this case, something like "self-managed object". Putting it on `QObject` would notify the tools that it manages memory automatically and no actions on the caller are necessary.

Or, yes, we could hardcode the name `QObject` in `MallocChecker.cpp` to escape upon construction.
